### PR TITLE
perf(dep_graph): Avoid allocating a set on when the number reads are …

### PR DESCRIPTION
**Pull request by Marwes**
…small

`reserve_and_rehash` takes up 1.4% of the runtime on the `packed-simd`
benchmark which I believe is due to the number of reads are very low in
many cases (see https://github.com/rust-lang/rust/pull/50565 for
instance).

This avoids allocating the set until we start allocating the `reads`
`SmallVec` but it is possible that a lower limit might be better (not
tested since the improvement will be hard to spot either way).

### CODE REVIEW

File: src/librustc/dep_graph/graph.rs [hunk 1]
```diff
@@ -1154,10 +1168,11 @@ impl DepGraphData {
     }
 }
 
+const TASK_DEPS_READS_CAP: usize = 8;
```
Comment (davidtwco) [inline src/librustc/dep_graph/graph.rs]: Could you add a doc comment on this?

**Review by davidtwco — COMMENTED** [review]
LGTM, one minor comment.

**Comment by Marwes** [comment]
An improvement, though it didn't affect `packed-simd` as I expected. It seems that the cost there is due to reallocting large sets.

**Review by davidtwco — APPROVED** [review]

**Comment by Zoxc** [comment]
This seems to [regress performance](https://perf.rust-lang.org/compare.html?start=865b44a3e330f3ef8be0f6edf69896c9ed957ac0&end=40a5ed489b28d2fffe246600ed04f0673d6ae7a2&stat=wall-time), but that could be perf noise.

**Pull request merged**